homeassistant/components/youtube/config_flow.py

```diff
@@ -0,0 +1,199 @@
+
+from collections.abc import Mapping
+import logging
+from typing import Any
+
+from google.oauth2.credentials import Credentials
+from googleapiclient.discovery import build
+from googleapiclient.http import HttpRequest
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, OptionsFlowWithConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.selector import (
+    SelectOptionDict,
+    SelectSelector,
+    SelectSelectorConfig,
+)
+
+from . import AsyncConfigEntryAuth
+from .const import AUTH, CONF_CHANNELS, DEFAULT_ACCESS, DOMAIN
+
+
+class OAuth2FlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=DOMAIN
+):
+    """Config flow to handle Google OAuth2 authentication."""
+
+    _data: dict[str, Any] = {}
+
+    DOMAIN = DOMAIN
+
+    reauth_entry: ConfigEntry | None = None
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: ConfigEntry,
+    ) -> YouTubeOptionsFlowHandler:
+        """Get the options flow for this handler."""
+        return YouTubeOptionsFlowHandler(config_entry)
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    @property
+    def extra_authorize_data(self) -> dict[str, Any]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {
+            "scope": " ".join(DEFAULT_ACCESS),
+            # Add params to ensure we get back a refresh token
+            "access_type": "offline",
+            "prompt": "consent",
+        }
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Perform reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm reauth dialog."""
+        if user_input is None:
+            return self.async_show_form(step_id="reauth_confirm")
+        return await self.async_step_user()
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> FlowResult:
+        """Create an entry for the flow, or update existing entry."""
+
+        service = build(
+            "youtube",
+            "v3",
+            credentials=Credentials(data[CONF_TOKEN][CONF_ACCESS_TOKEN]),
+        )
+        # pylint: disable=no-member
+        own_channel_request: HttpRequest = service.channels().list(
+            part="snippet", mine=True
+        )
+        response = await self.hass.async_add_executor_job(own_channel_request.execute)
```

Comment:
If i understand right, it seems like the channels list is queried twice during the config flow, so maybe the relevant details can be saved to avoid that.

Reply:
I want to update this logic to be more efficient in the future. A channel title and name don't update that frequent (the only data that we would miss would be an up to date subscriber count). So I want to fetch videos more frequent than channels, but to figure that out I wanted to open a PR later

Reply:
My understanding is its just getting the title later, so just save it now for later?

Reply:
You mean store all channel specific data in the entry, fetched in the config flow and then only update the latest video part?

Reply:
Yeah i'm just saying that `own_channel_request` is duplicate and i think it would be preferred to just issue that request once and save the data needed later (e.g. `self._title = ...`)

Reply:
OOOOH, I see what you mean, I forgot that I had to add it again for user id checks. I'll fix this